### DISCUSSION TITLE
Does react-query support Automatic persisted queries?

### DISCUSSION THREAD

**Discussion by mrchief** [TanStack/query#2230] [Q&A]
Something like this 

```
import { ApolloClient, InMemoryCache, HttpLink } from "@apollo/client";
import { createPersistedQueryLink } from "@apollo/client/link/persisted-queries";

const linkChain = createPersistedQueryLink().concat(new HttpLink({ uri: "http://localhost:4000/graphql" }));

const client = new ApolloClient({
  cache: new InMemoryCache(),
  link: linkChain,
});
```

> When the APQ link is created with createPersistedQueryLink({useGETForHashedQueries: true}), Apollo Client automatically sends the short hashed queries as GET requests allowing a CDN to serve those request.

https://www.apollographql.com/docs/apollo-server/performance/apq/

Does react-query have built-in mechanisms that do this or are there patterns to this with react-query?

**Comment by TkDodo** [comment]
This features seems to be specific to `apollo-server`. react-query is a promise based async state management library, so per design, it doesn't know or care how you do data fetching. You can also do `() => Promise.resolve(5)` as your "queryFn" and react-query is totally okay with it. That also means it cannot support any apollo-server magic out of the box :)

It seems to be easy to build in user-land though. Inspect the response, store the hash somewhere and send it for every future request instead of the body.

**Reply by mrchief** [reply]
My question was more about what apollo client does:

> This function creates a link that you can add to your client's Apollo Link chain. The link takes care of generating APQ identifiers, using GET requests for hashed queries, and retrying requests with query strings when necessary.

So generating the APQ, using GET for hashed etc. It seems like the answer is no as I cannot find any info anywhere about this but was wondering if anyone has explored this.

**Reply by dimified** [reply]
I am currently also looking into this topic and am interested in more findings on this.

**Reply by dimified** [reply]
Do you think it makes sense / Do you see any benefits in generating cacheable GET requests towards the GraphQL Service and use them combined with react-query or are staleTime and cacheTime options enough to meet the same purpose?

**Reply by mrchief** [reply]
GET queries with an APQ hash open up the possibility of caching the requests at Cloudfront or CDN nodes.

**Reply by dimified** [reply]
Thanks for the reply. Did you see any chance how to implement this with/without using Apollo Client and its APQ utilities?

**Reply by mrchief** [reply]
Not yet, still looking

**Reply by cromosom** [reply]
I think a middleware to translate hashed queries into post requests to the db and caching the responses for succeeding requests would be sufficient. Cache invalidation is still something to handle then

**Reply by binoiii** [reply]
By any chance, where you able to have an update on this @mrchief ?

**Reply by mrchief** [reply]
Unfortunately not. This seems like a complex thing and I didn't get enough time to spend on it.
